src/storage/src/render/upsert.rs

```diff
@@ -222,32 +295,104 @@ where
             // value2, time 7)
             let mut pending_values =
                 BTreeMap::<Timestamp, (Capability<Timestamp>, HashMap<_, UpsertSourceData>)>::new();
-            // this is a map of (decoded key) -> (decoded_value). We store the
-            // latest value for a given key that way we know what to retract if
-            // a new value with the same key comes along
-            let mut current_values = HashMap::new();
-
             // Intermediate structures re-used to limit allocations
             let mut scratch_vector = Vec::new();
+            let mut scratch_vector2 = Vec::new();
             let mut row_packer = mz_repr::Row::default();
             let mut dv = DatumVec::new();
 
-            // Prepare sorted and structured `key_indices` required
-            // by the upsert operator, and a `DatumVec` used to avoid
-            // an allocation.
-            let mut key_indices_sorted = upsert_envelope.key_indices.clone();
-            key_indices_sorted.sort_unstable();
             let key_indices_map = upsert_envelope
                 .key_indices
                 .iter()
                 .enumerate()
                 .map(|(idx, value_idx)| (*value_idx, idx))
                 .collect();
             let mut kdv = DatumVec::new();
+            // this is a map of (decoded key) -> (decoded_value). We store the
+            // latest value for a given key that way we know what to retract if
+            // a new value with the same key comes along
+            let mut current_values = None;
+
+            let mut initial_values_multiset: HashMap<(Row, Row), Diff> = HashMap::default();
+            move |data_input, previous_input, output| {
+                if previous_token.is_some() {
+                    assert!(current_values.is_none());
+                    // Hydrate the `current_values` map from the previous state of the collection.
+                    // We can't just insert things into the `current_values` map directly, since
+                    // we might in general have non-one multiplicities due to Persist being behind on compaction.
+                    // Thus, we use `initial_values_multiset` to keep track of how many of each record we've seen.
+                    //
+                    // At the end of reading the entire previous input, `initial_values_multiset` must have exactly one of each record,
+                    // and furthermore, each key must be unique. We validate this property for sanity's sake, and build `current_values`.
+                    //
+                    // TODO[btv] This algorithm has the potential to use unbounded space if we can't make any assumptions
+                    // about Persist's level of compaction or the order in which it returns updates.
+                    // See the detailed discussion [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1659063332027139).
```

Comment:
Idk what our norms are here, but a link to a Slack discussion isn't likely to be particularly timeless. Maybe that's fine, but if we end up with a filed issue that would be much better! Or, maybe everyone else disagrees.

Reply:
Filed an issue.